resourcex/extract.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcex
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+type ExtractOptions struct {
+	// TagName is the struct tag name to use for resource property names; defaults to "json".
+	TagName string
+	// RejectUnknowns produces an error (of type ContainsUnknownsError) if any unknowns are extracted.
+	RejectUnknowns bool
+}
+
+type ExtractResult struct {
+	// Dependencies is the set of resources that the extracted information depends on (known or unknown).
+	Dependencies []resource.URN
+	// ContainsUnknowns is true if the extracted information contains unknown values.
+	ContainsUnknowns bool
+	// ContainsSecrets is true if the extracted information contains secret values.
+	ContainsSecrets bool
+}
+
+// Extract extracts information from a property map into a target struct.
+// It returns a summary of the outputness and secretness of the extracted information.
+func Extract(target interface{}, props resource.PropertyMap, opts ExtractOptions) (ExtractResult, error) {
+	if opts.TagName == "" {
+		opts.TagName = "json"
+	}
+
+	// decode the property map into a JSON-like structure containing only values.
+	decoded := Decode(props)
+
+	// deserialize the JSON-like structure into a strongly typed struct.
+	config := &mapstructure.DecoderConfig{
```

Comment:
Does this strategy mean that we will never be able to support targeting `pulumix.OutputOf[T]`, `resource.Asset` and `resource.Archive`?

Reply:
The overarching idea is to create views of the values contained within a property map.  

Assets have a simple 'value' representation and may be decoded directly to `Asset`.  I implemented it in [a follow-up commit](https://github.com/pulumi/pulumi-go-provider/pull/191/commits/76bb7bdeaad8caff1cb763f503918e308f6d81a3).

Resource references are more interesting because `ResourceReference` has an `ID` field of type `PropertyValue`, which may be a string or be unknown.  In `Decode` one would emit an ID only if the value is known.  One could not extract a `ResourceReference` directly, but could extract a pure 'value' representation:
```
type ResourceReferenceValue struct {
  URN resource.URN `json:"urn,omitempty"`
  ID string `json:"id,omitempty"`
  PackageVersion string `json:"packageVersion,omitempty"`
}

Archives would require a similar 'value' representation.